compiler/rustc_span/src/symbol.rs

```diff
@@ -1512,9 +1512,12 @@ impl Ident {
         Ident::new(self.name, self.span.normalize_to_macro_rules())
     }
 
-    /// Convert the name to a `SymbolStr`. This is a slowish operation because
-    /// it requires locking the symbol interner.
-    pub fn as_str(self) -> SymbolStr {
+    /// Access the underlying string. This is a slowish operation because it
+    /// requires locking the symbol interner.
+    ///
+    /// Note that the lifetime of the return value is a lie. See
+    /// `Symbol::as_str()` for details.
```

Comment:
FWIW, the API was already unsound before this, the comment just makes it slightly more obvious.

Reply:
OTOH If anything this allows *less* code than the previous API (since the old return type outlived `'static`), so I am confused why new warnings about unsoundness are coming up now.

Reply:
Ah I found https://github.com/rust-lang/rust/pull/91957#discussion_r769812632, it was collapsed.

Reply:
well, we already had unsound functions in widely used rustc internal APIs. This PR just makes it less easy to abuse, so the PR by itself is not making the situation worse.

The only real way to fix this is to start tracking lifetimes for it everywhere, which is a much bigger change.

If we got rid of the use case of running a compiler twice in the same process we could probably just never drop the interner and then everything would be sound?

Reply:
> If we got rid of the use case of running a compiler twice in the same process we could probably just never drop the interner and then everything would be sound?

Wouldn't this break RLS?

Reply:
> the API was already unsound before this

Okay that doesn't really make me any happier but means it shouldn't be discussed in this PR:^^

Reply:
yes XD that's what I'm saying

Reply:
> This PR just makes it less easy to abuse,

Except for the `Send`/`Sync` difference.

Reply:
I don't get this argument. `as_str` now produces something that you can't really pass around at all, where `SymbolStr` had no lifetime binding it to `Symbol`. Sure you can send `Symbol` around, but you could do that before?

Reply:
The `&str` is `Send` and `Sync` either way.

Reply:
> The `&str` is `Send` and `Sync` either way.

Hm yeah that is fair.

Reply:
I think the only potential downside of the new API is that it makes it more annoying to unsoundly use it. Before, you would have had to store all the SymbolStrs somewhere, but now you just need the Symbols to come from a long-lived source.

Reply:
> I think the only potential downside of the new API is that it makes it more annoying to unsoundly use it

Is that a downside? Sounds like a good thing to me :smile: